beacon_node/beacon_chain/src/attestation_rewards.rs

```diff
@@ -0,0 +1,183 @@
+            let weight = get_flag_weight(flag_index as usize)
+                .map_err(|_| BeaconChainError::AttestationRewardsSyncError)?;
+
+            let unslashed_participating_indices = participation_cache
+                .get_unslashed_participating_indices(flag_index as usize, previous_epoch)?;
+
+            let unslashed_participating_balance =
+                unslashed_participating_indices.total_balance().unwrap();
+
+            let unslashed_participating_increments =
+                unslashed_participating_balance.safe_div(spec.effective_balance_increment)?;
+
+            let total_active_balance = participation_cache.current_epoch_total_active_balance();
+
+            let active_increments =
+                total_active_balance.safe_div(spec.effective_balance_increment)?;
+
+            let base_reward_per_increment =
+                BaseRewardPerIncrement::new(total_active_balance, spec)?;
+
+            for effective_balance_eth in 0..=32 {
+                let base_reward =
+                    effective_balance_eth.safe_mul(base_reward_per_increment.as_u64())?;
+
+                let penalty =
+                    !0 * base_reward.safe_mul(weight)?.safe_div(WEIGHT_DENOMINATOR)? as i64;
+
+                let reward_numerator = base_reward
+                    .safe_mul(weight)?
+                    .safe_mul(unslashed_participating_increments)?;
+
+                let ideal_reward = reward_numerator
+                    .safe_div(active_increments)?
+                    .safe_div(WEIGHT_DENOMINATOR)?;
+                if !state.is_in_inactivity_leak(previous_epoch, spec) {
+                    ideal_rewards_hashmap
+                        .insert((flag_index, effective_balance_eth, penalty), ideal_reward);
+                } else {
+                    ideal_rewards_hashmap.insert((flag_index, effective_balance_eth, penalty), 0);
+                }
+            }
+        }
+
+        // Calculate total rewards
+        let mut total_rewards: Vec<TotalAttestationRewards> = Vec::new();
+
+        if validators.is_empty() {
+            validators = participation_cache.eligible_validator_indices().to_vec();
+        }
+
+        for validator_index in &validators {
+            let eligible = state.is_eligible_validator(previous_epoch, *validator_index)?;
+
+            let effective_balance = state.get_effective_balance(*validator_index)?;
+
+            let effective_balance_eth =
+                effective_balance.safe_div(spec.effective_balance_increment)?;
+
+            let mut head_reward = 0u64;
+            let mut target_reward = 0i64;
+            let mut source_reward = 0i64;
+
+            for flag_index in PARTICIPATION_FLAG_WEIGHTS {
+                if eligible {
+                    let voted_correctly = participation_cache
+                        .get_unslashed_participating_indices(flag_index as usize, previous_epoch)
+                        .is_ok();
```

Comment:
We need to check that the specific `validator_index` voted correctly, at the moment this will always be `true` (assuming `get_unslashed_participating_indices` doesn't error).

```suggestion
                    let voted_correctly = participation_cache
                        .get_unslashed_participating_indices(flag_index as usize, previous_epoch)
                        .map_err(|_| Error::AttestationRewardsError)?
                        .contains(validator_index)
                        .map_err(|_| Error::AttestationRewardsError)?;
```